src/include/cart/api.h

```diff
@@ -706,227 +705,200 @@ crt_ep_abort(crt_endpoint_t *ep);
-		CRT_GEN_GET_TYPE(seq))					\
-	BOOST_PP_IF(BOOST_PP_EQUAL(CRT_GEN_X CRT_PTR, CRT_GEN_GET_KIND(seq)), \
-		*CRT_GEN_GET_NAME(seq), CRT_GEN_GET_NAME(seq));)
-
-#define CRT_GEN_STRUCT_FIELDS(z, n, seq)				\
-	CRT_GEN_STRUCT_FIELD(BOOST_PP_SEQ_ELEM(n, seq))
-
-#define CRT_GEN_STRUCT(struct_type_name, seq)				\
-	struct struct_type_name {					\
-		BOOST_PP_REPEAT(BOOST_PP_SEQ_SIZE(seq),			\
-				CRT_GEN_STRUCT_FIELDS, seq)		\
+#define CRT_GEN_STRUCT_FIELD(seq)                                                                  \
+	BOOST_PP_EXPAND(BOOST_PP_IF(                                                               \
+			    BOOST_PP_EQUAL(CRT_GEN_X CRT_ARRAY, CRT_GEN_GET_KIND(seq)),            \
+			    struct {                                                               \
+				    uint64_t ca_count;                                             \
+				    CRT_GEN_GET_TYPE(seq) * ca_arrays;                             \
+			    },                                                                     \
+			    CRT_GEN_GET_TYPE(seq))                                                 \
+			    BOOST_PP_IF(BOOST_PP_EQUAL(CRT_GEN_X CRT_PTR, CRT_GEN_GET_KIND(seq)),  \
+					*CRT_GEN_GET_NAME(seq), CRT_GEN_GET_NAME(seq));)
+
+#define CRT_GEN_STRUCT_FIELDS(z, n, seq) CRT_GEN_STRUCT_FIELD(BOOST_PP_SEQ_ELEM(n, seq))
+
+#define CRT_GEN_STRUCT(struct_type_name, seq)                                                      \
+	struct struct_type_name {                                                                  \
+		BOOST_PP_REPEAT(BOOST_PP_SEQ_SIZE(seq), CRT_GEN_STRUCT_FIELDS, seq)                \
 	};
 
-#define CRT_GEN_PROC_ARRAY(ptr, seq)					\
-	do {								\
-	CRT_GEN_GET_TYPE(seq) **e_ptrp = &ptr->CRT_GEN_GET_NAME(seq).ca_arrays;\
-	CRT_GEN_GET_TYPE(seq) *e_ptr = ptr->CRT_GEN_GET_NAME(seq).ca_arrays; \
-	uint64_t count = ptr->CRT_GEN_GET_NAME(seq).ca_count;		\
-	uint64_t i;							\
-	if (proc_op == CRT_PROC_DECODE)					\
-		*e_ptrp = NULL;						\
-	/* process the count of array first */				\
-	rc = crt_proc_uint64_t(proc, proc_op, &count);			\
-	if (unlikely(rc))						\
-		goto CRT_GEN_FAIL_LABEL(seq);				\
-	ptr->CRT_GEN_GET_NAME(seq).ca_count = count;			\
-	if (unlikely(count == 0))					\
-		break; /* goto next field */				\
-	if (proc_op == CRT_PROC_DECODE) {				\
-		D_ALLOC_ARRAY(e_ptr, (int)count);			\
-		if (unlikely(e_ptr == NULL)) {				\
-			rc = -DER_NOMEM;				\
-			goto CRT_GEN_FAIL_LABEL(seq);			\
-		}							\
-		*e_ptrp = e_ptr;					\
-	}								\
-	/* process the elements of array */				\
-	for (i = 0; i < count && e_ptr != NULL; i++) {			\
-		rc = CRT_GEN_GET_FUNC(seq)(proc, proc_op, &e_ptr[i]);	\
-		if (unlikely(rc)) {					\
-			if (proc_op == CRT_PROC_DECODE)			\
-				ptr->CRT_GEN_GET_NAME(seq).ca_count = i;\
-			goto CRT_GEN_FAIL_LABEL(seq);			\
-		}							\
-	}								\
-	if (proc_op == CRT_PROC_FREE)					\
-		D_FREE(*e_ptrp);					\
+#define CRT_GEN_PROC_ARRAY(ptr, seq)                                                               \
```

Comment:
(style)  Macros with flow control statements should be avoided
(style)  do {} while (0) macros should not be semicolon terminated